pkg/pubsub/system.go

```diff
@@ -0,0 +1,89 @@
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/connection"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/provider"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("pubsub-system")
+
+type System struct {
+	mux         sync.RWMutex
+	connections map[string]connection.Connection
+}
+
+func NewSystem() *System {
+	return &System{}
+}
+
+func (p *System) Publish(ctx context.Context, connection string, topic string, msg interface{}) error {
+	p.mux.RLock()
+	defer p.mux.RUnlock()
+	if len(p.connections) > 0 {
+		if c, ok := p.connections[connection]; ok {
+			return c.Publish(context.Background(), msg, topic)
+		}
+		log.Info("Connection is not initialized", "name", connection)
+	} else {
+		log.Info("No pub sub tools are enabled")
+	}
+	return nil
+}
+
+func (p *System) PublishBatch(ctx context.Context, connection string, topic string, msg []interface{}) error {
```

Comment:
This is where the "emulate batch publishing" code should go. Roughly:

```golang
connection := p.connections[connectionName]
if connection.SupportsBatching() { // `supportsBatching()` is probably a more descriptive name
  return connection.PublishBatch(ctx, topic, msgs)
}

// if still executing, batching is not natively supported by the connection, emulate it
errs := []error{} // initialize some bulk error handler
for i := range msgs {
  if err := connection.Publish(ctx, topic, msg); err != nil {
    // figure out how error handling works, particularly if there is partial success
   // should there be retries?
  }
}
```

Note in the above snippet I'm only including logic directly related to emulating batch (and only the rough shape of it at that), not attempting to have proper locking, error handling, etc.

Reply:
I am wondering why are we taking responsibility for implementing batch publishing ourselves if some provider doesn't support it. My argument is we should just support whatever provider supports, otherwise, we will be implementing a lot more for batch publishing. For example the handling of partial success, retries logic, etc. Rather, we should depend on the driver and by extension respective messaging tool to handle such complications. If we want to write a simple driver/interface then that should be limited to what the underlying broker supports right? I am wondering if we are overcomplicating and over-engineering some of the things or not.